cpp/dspace/doi.cc

```diff
@@ -0,0 +1,98 @@
+#include "HttpHeader.h"
+#include "JSON.h"
+#include "StlHelpers.h"
+#include "StringUtil.h"
+#include "VuFind.h"
+#include "util.h"
+
+using namespace std;
+
+const std::string DspaceServerURL("https://publikationen.uni-tuebingen.de/rest/items/");
+const std::string DOIKey("dc.relation.uri");
+const std::string PublicationTitleKey("dc.title");
+const std::vector<std::string> notification_email_addresses = { "andrii.lysohor@uni-tuebingen.de" };
+
+void SendNotificationsForDOI(const std::string email_subject, const std::string email_message) {
+    if (EmailSender::SimplerSendEmail("no-reply@ub.uni-tuebingen.de", notification_email_addresses, email_subject, email_message,
+                                      EmailSender::VERY_HIGH)
+        > 299)
+        LOG_ERROR("Failed to send the DOI notification email!");
+}
+
+void UpdateItem(DbConnection * const db_writer, const std::string doi_link, const std::string publication_id) {
+    db_writer->queryOrDie("UPDATE tuefind_publications SET doi_link = " + db_writer->escapeAndQuoteString(doi_link)
+                          + ",doi_notification = CURDATE() WHERE id= " + publication_id + "");
+}
+
+bool DownloadAndUpdate(DbConnection * const db_writer, const std::string external_document_id, const std::string publicationID) {
+    const std::string DOWNLOAD_URL(DspaceServerURL + external_document_id + "/metadata");
+
+    Downloader downloader(DOWNLOAD_URL, Downloader::Params(), 15 * 1000);
+    if (downloader.anErrorOccurred()) {
+        LOG_WARNING("Error while downloading data for id : " + downloader.getLastErrorMessage());
+        return false;
+    }
+
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() != 200) {
+        LOG_WARNING("DOI returned HTTP status code " + std::to_string(http_header.getStatusCode()) + "! for item id: " + publicationID);
+        return false;
+    }
+
+    const std::string &json_document(downloader.getMessageBody());
+    std::shared_ptr<JSON::JSONNode> full_tree;
+    JSON::Parser parser(json_document);
+    if (not parser.parse(&full_tree))
+        LOG_ERROR("failed to parse JSON (" + parser.getErrorMessage());
+
+    const std::shared_ptr<const JSON::ArrayNode> top_node_array(JSON::JSONNode::CastToArrayNodeOrDie("full_tree", full_tree));
+
+    std::string PublicationTitle;
+
+    for (auto item_iter : *top_node_array) {
+        const std::string title_key(JSON::LookupString("/key", item_iter, ""));
+        const std::string title_value(JSON::LookupString("/value", item_iter, ""));
+        if (PublicationTitleKey == title_key) {
+            PublicationTitle = title_value;
+        }
+    }
+
+    for (auto item_iter : *top_node_array) {
+        const std::string one_key(JSON::LookupString("/key", item_iter, ""));
+        const std::string DOILink(JSON::LookupString("/value", item_iter, ""));
+        if (DOIKey == one_key) {
+            std::cout << "Processing: " << DOILink << '\n';
```

Comment:
Usually we should not write to std::cout directly. Please use `LOG_INFO(doi_link)` instead.